sdk/python/src/dagger/server/_server.py

```diff
@@ -0,0 +1,116 @@
+import inspect
+import logging
+from collections.abc import Callable
+from functools import partial
+from pathlib import Path
+from typing import Any, cast
+
+import anyio
+import attrs
+from cattrs.preconf.json import JsonConverter
+from strawberry import Schema
+from strawberry.utils.await_maybe import await_maybe
+
+from ._context import Context
+from ._converter import converter as json_converter
+from ._converter import register_dagger_type_hooks
+
+logger = logging.getLogger(__name__)
+
+inputs_path = Path("/inputs/dagger.json")
+outputs_path = Path("/outputs/dagger.json")
+schema_path = Path("/outputs/schema.graphql")
+
+
+@attrs.define
+class Inputs:
+    resolver: str
+    args: dict[str, Any]
+    parent: dict[str, Any] | None
+
+
+@attrs.define
+class Server:
+    schema: Schema
+    converter: JsonConverter = json_converter
+    debug: bool = False
+
+    def export_schema(self) -> None:
+        logger.debug("schema => \n%s", self.schema)
+        schema_path.write_text(str(self.schema))
+
+    def execute(self) -> None:
+        inputs = self._read_inputs()
+        logger.debug("inputs = %s", inputs)
+
+        output = anyio.run(self._call_resolver, inputs)
+
+        logger.debug("output = %s", output)
+        self._write_output(output)
+
+    def _read_inputs(self) -> Inputs:
+        return self.converter.loads(inputs_path.read_text(), Inputs)
+
+    def _write_output(self, out: str) -> None:
+        outputs_path.write_text(out)
+
+    async def _call_resolver(self, inputs: Inputs) -> str:
+        type_name, field_name = inputs.resolver.split(".", 2)
+        field = self.schema.get_field_for_type(field_name, type_name)
+        if field is None:
+            # FIXME: use proper error class
+            msg = f"Can't find field `{field_name}` for type `{type_name}`"
+            raise ValueError(msg)
+
+        resolver: Callable = self.schema.schema_converter.from_resolver(field)
+
+        # origin is the parent type of the resolver/field.
+        origin = cast(Callable, field.origin)
+
+        async with Context() as context:
+            register_dagger_type_hooks(self.converter, context)
+
+            # inputs.parent is a dict of the parent type's fields.
+            if inputs.parent is None:
+                parent = origin()
+            elif inspect.isclass(origin):
+                parent = await anyio.to_thread.run_sync(
+                    self.converter.structure,
+                    inputs.parent,
+                    origin,
+                )
+            else:
+                # TODO: When is origin not a class?
```

Comment:
IIUC, the parent type should always be a struct of some sort, so curious if you were hitting a situation where it wasn't.

Reply:
No, no issue, just my brain being thorough with the if branching. 🙂 My initial impulse was to raise an exception in the `else` branch, but then wondered how to test it.

I had a simple `origin(**inputs.parent) if inputs.parent else origin()` previously and the converter stuff is only needed when it's a class, although it should work with any type.

According to the type I get from `origin`, it can be a normal function (which isn't a type I can convert to). It's not clear to me yet when that's the case, as I haven't given it much thought. I don't think it's something we need in this opinionated layer on top of strawberry, but this module should remain generic even for full extension support.